File: gaia-sdk-javascript/src/api/ISensorFunction.ts
import {Observable} from "rxjs";
import {
    BehaviourReq,
    BehaviourRes,
    CodeReq,
    CodeRes,
    CreatedIntentImpulse,
    CreateIntentImpulse,
    DeletedIntentImpulse,
    DeleteIntentImpulse,
    EdgeReq,
    EdgeRes,
    ExperienceReq,
    ExperienceRes,
    FulfilmentReq,
    FulfilmentRes,
    IdentityReq,
    IdentityRes,
    TenantReq,
    TenantRes,
    UserReq,
    UserRes,
    IntentReq,
    IntentRes,
    IntrospectionReq,
    IntrospectionRes,
    KnowledgeReq,
    KnowledgeRes,
    PerceiveActionImpulse,
    PerceiveDataImpulse,
    PerceivedImpulse,
    PerceptionReq,
    PerceptionRes,
    PreservationReq,
    PreservationRes,
    PromptReq,
    PromptRes,
    RetrievalReq,
    RetrievalRes,
    SkillIntrospectionReq,
    SkillIntrospectionRes,
    SkillProvisionReq,
    SkillProvisionRes,
    SkillReq,
    SkillRes,
    BehaviourExecutionReq,
    BehaviourExecutionRes,
    BehaviourExecutionDetailReq,
    BehaviourExecutionDetailRes,
    MetricsReq,
    MetricsRes,
    StatementReq,
    StatementRes,
    UpdatedIntentImpulse,
    UpdateIntentImpulse, ApiKeyReq, ApiKeyRes, RoleReq, RoleRes, SkillProvisionBuildJobReq
} from "../graphql";
import {CreatePromptImpulse} from "../graphql/request/input/CreatePromptImpulse";
import {CreatedPromptImpulse} from "../graphql/response/type/CreatedPromptImpulse";
import {UpdatePromptImpulse} from "../graphql/request/input/UpdatePromptImpulse";
import {UpdatedPromptImpulse} from "../graphql/response/type/UpdatedPromptImpulse";
import {DeletePromptImpulse} from "../graphql/request/input/DeletePromptImpulse";
import {DeletedPromptImpulse} from "../graphql/response/type/DeletedPromptImpulse";
import {CreateStatementImpulse} from "../graphql/request/input/CreateStatementImpulse";
import {CreatedStatementImpulse} from "../graphql/response/type/CreatedStatementImpulse";
import {UpdateStatementImpulse} from "../graphql/request/input/UpdateStatementImpulse";
import {UpdatedStatementImpulse} from "../graphql/response/type/UpdatedStatementImpulse";
import {DeleteStatementImpulse} from "../graphql/request/input/DeleteStatementImpulse";
import {DeletedStatementImpulse} from "../graphql/response/type/DeletedStatementImpulse";
import {CreateFulfilmentImpulse} from "../graphql/request/input/CreateFulfilmentImpulse";
import {CreatedFulfilmentImpulse} from "../graphql/response/type/CreatedFulfilmentImpulse";
import {UpdateFulfilmentImpulse} from "../graphql/request/input/UpdateFulfilmentImpulse";
import {UpdatedFulfilmentImpulse} from "../graphql/response/type/UpdatedFulfilmentImpulse";
import {DeleteFulfilmentImpulse} from "../graphql/request/input/DeleteFulfilmentImpulse";
import {DeletedFulfilmentImpulse} from "../graphql/response/type/DeletedFulfilmentImpulse";
import {CreateBehaviourImpulse} from "../graphql/request/input/CreateBehaviourImpulse";
import {CreatedBehaviourImpulse} from "../graphql/response/type/CreatedBehaviourImpulse";
import {UpdateBehaviourImpulse} from "../graphql/request/input/UpdateBehaviourImpulse";
import {UpdatedBehaviourImpulse} from "../graphql/response/type/UpdatedBehaviourImpulse";
import {DeleteBehaviourImpulse} from "../graphql/request/input/DeleteBehaviourImpulse";
import {DeletedBehaviourImpulse} from "../graphql/response/type/DeletedBehaviourImpulse";
import {CreateCodeImpulse} from "../graphql/request/input/CreateCodeImpulse";
import {CreatedCodeImpulse} from "../graphql/response/type/CreatedCodeImpulse";
import {UpdateCodeImpulse} from "../graphql/request/input/UpdateCodeImpulse";
import {UpdatedCodeImpulse} from "../graphql/response/type/UpdatedCodeImpulse";
import {DeleteCodeImpulse} from "../graphql/request/input/DeleteCodeImpulse";
import {DeletedCodeImpulse} from "../graphql/response/type/DeletedCodeImpulse";
import {CreateEdgeImpulse} from "../graphql/request/input/CreateEdgeImpulse";
import {CreatedEdgeImpulse} from "../graphql/response/type/CreatedEdgeImpulse";
import {DeleteEdgeImpulse} from "../graphql/request/input/DeleteEdgeImpulse";
import {DeletedEdgeImpulse} from "../graphql/response/type/DeletedEdgeImpulse";
import {CreateSkillImpulse} from "../graphql/request/input/CreateSkillImpulse";
import {CreatedSkillImpulse} from "../graphql/response/type/CreatedSkillImpulse";
import {UpdateSkillImpulse} from "../graphql/request/input/UpdateSkillImpulse";
import {UpdatedSkillImpulse} from "../graphql/response/type/UpdatedSkillImpulse";
import {DeleteSkillImpulse} from "../graphql/request/input/DeleteSkillImpulse";
import {DeletedSkillImpulse} from "../graphql/response/type/DeletedSkillImpulse";
import {CreateSkillProvisionImpulse} from "../graphql/request/input/CreateSkillProvisionImpulse";
import {CreatedSkillProvisionImpulse} from "../graphql/response/type/CreatedSkillProvisionImpulse";
import {UpdateSkillProvisionImpulse} from "../graphql/request/input/UpdateSkillProvisionImpulse";
import {UpdatedSkillProvisionImpulse} from "../graphql/response/type/UpdatedSkillProvisionImpulse";
import {DeleteSkillProvisionImpulse} from "../graphql/request/input/DeleteSkillProvisionImpulse";
import {DeletedSkillProvisionImpulse} from "../graphql/response/type/DeletedSkillProvisionImpulse";
import {Struct, Uuid} from "../graphql/GaiaClient";
import {CreateIdentityImpulse} from "../graphql/request/input/CreateIdentityImpulse";
import {UpdateIdentityImpulse} from "../graphql/request/input/UpdateIdentityImpulse";
import {DeleteIdentityImpulse} from "../graphql/request/input/DeleteIdentityImpulse";
import {CreatedIdentityImpulse} from "../graphql/response/type/CreatedIdentityImpulse";
import {UpdatedIdentityImpulse} from "../graphql/response/type/UpdatedIdentityImpulse";
import {DeletedIdentityImpulse} from "../graphql/response/type/DeletedIdentityImpulse";
import {CreateTenantImpulse} from "../graphql/request/input/CreateTenantImpulse";
import {UpdateTenantImpulse} from "../graphql/request/input/UpdateTenantImpulse";
import {DeleteTenantImpulse} from "../graphql/request/input/DeleteTenantImpulse";
import {CreatedTenantImpulse} from "../graphql/response/type/CreatedTenantImpulse";
import {UpdatedTenantImpulse} from "../graphql/response/type/UpdatedTenantImpulse";
import {DeletedTenantImpulse} from "../graphql/response/type/DeletedTenantImpulse";
import {CreateUserImpulse} from "../graphql/request/input/CreateUserImpulse";
import {UpdateUserImpulse} from "../graphql/request/input/UpdateUserImpulse";
import {DeleteUserImpulse} from "../graphql/request/input/DeleteUserImpulse";
import {UpdatedUserImpulse} from "../graphql/response/type/UpdatedUserImpulse";
import {CreatedUserImpulse} from "../graphql/response/type/CreatedUserImpulse";
import {DeletedUserImpulse} from "../graphql/response/type/DeletedUserImpulse";
import {CreateApiKeyImpulse} from "../graphql/request/input/CreateApiKeyImpulse";
import {CreatedApiKeyImpulse} from "../graphql/response/type/CreatedApiKeyImpulse";
import {UpdateApiKeyImpulse} from "../graphql/request/input/UpdateApiKeyImpulse";
import {UpdatedApiKeyImpulse} from "../graphql/response/type/UpdatedApiKeyImpulse";
import {DeleteApiKeyImpulse} from "../graphql/request/input/DeleteApiKeyImpulse";
import {DeletedApiKeyImpulse} from "../graphql/response/type/DeletedApiKeyImpulse";
import {CreateRoleImpulse} from "../graphql/request/input/CreateRoleImpulse";
import {CreatedRoleImpulse} from "../graphql/response/type/CreatedRoleImpulse";
import {UpdateRoleImpulse} from "../graphql/request/input/UpdateRoleImpulse";
import {UpdatedRoleImpulse} from "../graphql/response/type/UpdatedRoleImpulse";
import {DeleteRoleImpulse} from "../graphql/request/input/DeleteRoleImpulse";
import {DeletedRoleImpulse} from "../graphql/response/type/DeletedRoleImpulse";
import {EdgeType} from "../graphql/request/enumeration/EdgeType";
import {ConnectNodeSetImpulse} from "../graphql/response/type/ConnectNodeSetImpulse";
import {ConnectNodeUnsetImpulse} from "../graphql/response/type/ConnectNodeUnsetImpulse";
import {ConnectNodeAppendedImpulse} from "../graphql/response/type/ConnectNodeAppendedImpulse";
import {ConnectNodeRemovedImpulse} from "../graphql/response/type/ConnectNodeRemovedImpulse";
import {ConnectSetNodeImpulse} from "../graphql/request/input/ConnectSetNodeImpulse";
import {ConnectAppendNodeImpulse} from "../graphql/request/input/ConnectAppendNodeImpulse";
import {ConnectRemoveNodeImpulse} from "../graphql/request/input/ConnectRemoveNodeImpulse";
import {ConnectUnsetNodeImpulse} from "../graphql/request/input/ConnectUnsetNodeImpulse";
import {BehaviourExecution} from "../graphql/request/type/BehaviourExecution";

export interface ISensorFunction {
    retrieve(config: (x: RetrievalReq) => void): Observable<RetrievalRes>

    retrieveExperience(config: (x: ExperienceReq) => void): Observable<ExperienceRes>

    retrieveKnowledge(config: (x: KnowledgeReq) => void): Observable<KnowledgeRes>

    retrieveEdges(source: Uuid, config: (x: EdgeReq) => void, limit?: Number, offset?: Number): Observable<EdgeRes>

    retrieveEdge(source: Uuid, edgeId: Uuid, config: (x: EdgeReq) => void): Observable<EdgeRes>

    retrieveIdentities(config: (x: IdentityReq) => void, limit?: Number, offset?: Number): Observable<IdentityRes>

    retrieveIdentity(identityId: Uuid, config: (x: IdentityReq) => void): Observable<IdentityRes>

    retrieveTenants(config: (x: TenantReq) => void, limit?: Number, offset?: Number): Observable<TenantRes>

    retrieveTenant(tenantId: Uuid, config: (x: TenantReq) => void) : Observable<TenantRes>

    retrieveUsers(config: (x: UserReq) => void, limit?: Number, offset?: Number): Observable<UserRes>

    retrieveUser(userId: Uuid, config: (x: UserReq) => void) : Observable<UserRes>

    retrieveApiKeys(config: (x: ApiKeyReq) => void, limit?: Number, offset?: Number): Observable<ApiKeyRes>

    retrieveApiKey(apiKeyId: Uuid, config: (x: ApiKeyReq) => void) : Observable<ApiKeyRes>

    retrieveRoles(tenantId: Uuid, config: (x: RoleReq) => void, limit?: Number, offset?: Number): Observable<RoleRes>

    retrieveRole(tenantId: Uuid, roleId: Uuid, config: (x: RoleReq) => void) : Observable<RoleRes>

    retrieveIntents(identityId: Uuid, config: (x: IntentReq) => void, limit?: Number, offset?: Number): Observable<IntentRes>

    retrieveIntent(identityId: Uuid, reference, config: (x: IntentReq) => void): Observable<IntentRes>

    retrievePrompts(identityId: Uuid, config: (x: PromptReq) => void, limit?: Number, offset?: Number): Observable<PromptRes>

    retrievePrompt(identityId: Uuid, reference: Uuid, config: (x: PromptReq) => void): Observable<PromptRes>

    retrieveStatements(identityId: Uuid, config: (x: StatementReq) => void, limit?: Number, offset?: Number): Observable<StatementRes>

    retrieveStatement(identityId: Uuid, reference: Uuid, config: (x: StatementReq) => void): Observable<StatementRes>

    retrieveFulfilments(identityId: Uuid, config: (x: FulfilmentReq) => void, limit?: Number, offset?: Number): Observable<FulfilmentRes>

    retrieveFulfilment(identityId: Uuid, reference: Uuid, config: (x: FulfilmentReq) => void): Observable<FulfilmentRes>

    retrieveCodes(identityId: Uuid, config: (x: CodeReq) => void, limit?: Number, offset?: Number): Observable<CodeRes>

    retrieveCode(identityId: Uuid, reference: Uuid, config: (x: CodeReq) => void): Observable<CodeRes>

    retrieveBehaviours(identityId: Uuid, config: (x: BehaviourReq) => void, limit?: Number, offset?: Number): Observable<BehaviourRes>

    retrieveBehaviour(identityId: Uuid, reference: Uuid, config: (x: BehaviourReq) => void): Observable<BehaviourRes>

    retrieveSkills(tenantId: Uuid, config: (x: SkillReq) => void, limit?: Number, offset?: Number): Observable<SkillRes>

    retrieveSkill(tenantId: Uuid, reference: Uuid, config: (x: SkillReq) => void): Observable<SkillRes>

    retrieveSkillProvisions(tenantId: Uuid, config: (x: SkillProvisionReq) => void, limit?: Number, offset?: Number): Observable<SkillProvisionRes>

    retrieveSkillProvision(tenantId: Uuid, reference: Uuid, config: (x: SkillProvisionReq) => void): Observable<SkillProvisionRes>

    retrieveSkillProvisionBuildJobs(tenantId: Uuid, config: (x: SkillProvisionBuildJobReq) => void): Observable<SkillIntrospectionRes>

    retrieveBehaviourExecution(identityId: Uuid, processInstanceId: Uuid, config: (x: BehaviourExecutionDetailReq) => void): Observable<BehaviourExecutionDetailRes>

    retrieveBehaviourExecutions(identityId: Uuid, config: (x: BehaviourExecutionReq) => void, limit?: Number, offset?: Number, startDate?: string, endDate?: string): Observable<BehaviourExecutionRes>

    retrieveMetrics(identityId: Uuid, config: (x: MetricsReq) => void): Observable<MetricsRes>

    introspect(config: (x: IntrospectionReq) => void): Observable<IntrospectionRes>

    introspectSkills(config: (x: SkillIntrospectionReq) => void): Observable<SkillIntrospectionRes>

    preserve(config: (x: PreservationReq) => void): Observable<PreservationRes>

    preserveCreateIdentities(...impulses: [CreateIdentityImpulse]): Observable<CreatedIdentityImpulse>

    preserveUpdateIdentities(...impulses: [UpdateIdentityImpulse]): Observable<UpdatedIdentityImpulse>

    preserveDeleteIdentities(...impulses: [DeleteIdentityImpulse]): Observable<DeletedIdentityImpulse>

    preserveCreateTenants(...impulses: [CreateTenantImpulse]): Observable<CreatedTenantImpulse>

    preserveUpdateTenants(...impulses: [UpdateTenantImpulse]): Observable<UpdatedTenantImpulse>

    preserveDeleteTenants(...impulses: [DeleteTenantImpulse]): Observable<DeletedTenantImpulse>

    preserveCreateUsers(...impulses: [CreateUserImpulse]): Observable<CreatedUserImpulse>

    preserveUpdateUsers(...impulses: [UpdateUserImpulse]): Observable<UpdatedUserImpulse>

    preserveDeleteUsers(...impulses: [DeleteUserImpulse]): Observable<DeletedUserImpulse>

    preserveCreateApiKeys(...impulses: [CreateApiKeyImpulse]): Observable<CreatedApiKeyImpulse>

    preserveUpdateApiKeys(...impulses: [UpdateApiKeyImpulse]): Observable<UpdatedApiKeyImpulse>

    preserveDeleteApiKeys(...impulses: [DeleteApiKeyImpulse]): Observable<DeletedApiKeyImpulse>

    preserveCreateRoles(...impulses: [CreateRoleImpulse]): Observable<CreatedRoleImpulse>

    preserveUpdateRoles(...impulses: [UpdateRoleImpulse]): Observable<UpdatedRoleImpulse>

    preserveDeleteRoles(...impulses: [DeleteRoleImpulse]): Observable<DeletedRoleImpulse>

    preserveCreateIntents(...impulses: [CreateIntentImpulse]): Observable<CreatedIntentImpulse>

    preserveUpdateIntents(...impulses: [UpdateIntentImpulse]): Observable<UpdatedIntentImpulse>

    preserveDeleteIntents(...impulses: [DeleteIntentImpulse]): Observable<DeletedIntentImpulse>

    preserveCreatePrompts(...impulses: [CreatePromptImpulse]): Observable<CreatedPromptImpulse>

    preserveUpdatePrompts(...impulses: [UpdatePromptImpulse]): Observable<UpdatedPromptImpulse>

    preserveDeletePrompts(...impulses: [DeletePromptImpulse]): Observable<DeletedPromptImpulse>

    preserveCreateStatements(...impulses: [CreateStatementImpulse]): Observable<CreatedStatementImpulse>

    preserveUpdateStatements(...impulses: [UpdateStatementImpulse]): Observable<UpdatedStatementImpulse>

    preserveDeleteStatements(...impulses: [DeleteStatementImpulse]): Observable<DeletedStatementImpulse>

    preserveCreateFulfilments(...impulses: [CreateFulfilmentImpulse]): Observable<CreatedFulfilmentImpulse>

    preserveUpdateFulfilments(...impulses: [UpdateFulfilmentImpulse]): Observable<UpdatedFulfilmentImpulse>

    preserveDeleteFulfilments(...impulses: [DeleteFulfilmentImpulse]): Observable<DeletedFulfilmentImpulse>

    preserveCreateBehaviours(...impulses: [CreateBehaviourImpulse]): Observable<CreatedBehaviourImpulse>

    preserveUpdateBehaviours(...impulses: [UpdateBehaviourImpulse]): Observable<UpdatedBehaviourImpulse>

    preserveDeleteBehaviours(...impulses: [DeleteBehaviourImpulse]): Observable<DeletedBehaviourImpulse>

    preserveCreateCodes(...impulses: [CreateCodeImpulse]): Observable<CreatedCodeImpulse>

    preserveUpdateCodes(...impulses: [UpdateCodeImpulse]): Observable<UpdatedCodeImpulse>

    preserveDeleteCodes(...impulses: [DeleteCodeImpulse]): Observable<DeletedCodeImpulse>

    preserveCreateEdges(...impulses: [CreateEdgeImpulse]): Observable<CreatedEdgeImpulse>

    preserveDeleteEdges(...impulses: [DeleteEdgeImpulse]): Observable<DeletedEdgeImpulse>

    preserveConnectNodeSet(nodeId: Uuid, impulse: ConnectSetNodeImpulse): Observable<ConnectNodeSetImpulse>

    preserveConnectNodeUnset(nodeId: Uuid, impulse: ConnectUnsetNodeImpulse): Observable<ConnectNodeUnsetImpulse>

    preserveConnectNodeAppend(nodeId: Uuid, impulse: ConnectAppendNodeImpulse): Observable<ConnectNodeAppendedImpulse>

    preserveConnectNodeRemove(nodeId: Uuid, impulse: ConnectRemoveNodeImpulse): Observable<ConnectNodeRemovedImpulse>

    preserveCreateSkills(...impulses: [CreateSkillImpulse]): Observable<CreatedSkillImpulse>

    preserveUpdateSkills(...impulses: [UpdateSkillImpulse]): Observable<UpdatedSkillImpulse>

    preserveDeleteSkills(...impulses: [DeleteSkillImpulse]): Observable<DeletedSkillImpulse>

    preserveCreateSkillProvisions(...impulses: [CreateSkillProvisionImpulse]): Observable<CreatedSkillProvisionImpulse>

    preserveUpdateSkillProvisions(...impulses: [UpdateSkillProvisionImpulse]): Observable<UpdatedSkillProvisionImpulse>

    preserveDeleteSkillProvisions(...impulses: [DeleteSkillProvisionImpulse]): Observable<DeletedSkillProvisionImpulse>

    perceive(config: (x: PerceptionReq) => void): Observable<PerceptionRes>

    perceiveAction(impulse: PerceiveActionImpulse): Observable<PerceivedImpulse>

    perceiveData(impulse: PerceiveDataImpulse): Observable<PerceivedImpulse>
}

File: gaia-sdk-python/gaia_sdk/api/ISensorFunction.py
from abc import abstractmethod, ABC
from typing import Callable, List, Union

from rx.core.typing import Observable

from gaia_sdk.graphql import RetrievalReq, ExperienceReq, KnowledgeReq, EdgeReq, \
    IdentityReq, IntentReq, PromptReq, StatementReq, FulfilmentReq, CodeReq, BehaviourReq, IntrospectionReq, \
    PerceptionReq, PreservationReq, CreatedIdentityImpulse, UpdatedIdentityImpulse, DeletedIdentityImpulse, \
    CreatedIntentImpulse, UpdatedIntentImpulse, DeletedIntentImpulse, RetrievalRes, \
    ExperienceRes, EdgeRes, StatementRes, PromptRes, IntentRes, IdentityRes, KnowledgeRes, FulfilmentRes, CodeRes, \
    BehaviourRes, IntrospectionRes, PreservationRes, PerceptionRes, PerceivedImpulse, \
    DeleteIdentityImpulse, UpdateIdentityImpulse, CreateIdentityImpulse, DeleteIntentImpulse, UpdateIntentImpulse, \
    CreateIntentImpulse, \
    PerceiveActionImpulse, PerceiveDataImpulse, \
    CreatePromptImpulse, UpdatePromptImpulse, DeletePromptImpulse, CreatedPromptImpulse, UpdatedPromptImpulse, \
    DeletedPromptImpulse, CreateStatementImpulse, UpdateStatementImpulse, DeleteStatementImpulse, \
    CreatedStatementImpulse, UpdatedStatementImpulse, DeletedStatementImpulse, CreateFulfilmentImpulse, \
    UpdateFulfilmentImpulse, DeleteFulfilmentImpulse, CreatedFulfilmentImpulse, UpdatedFulfilmentImpulse, \
    DeletedFulfilmentImpulse, CreateBehaviourImpulse, UpdateBehaviourImpulse, DeleteBehaviourImpulse, \
    CreatedBehaviourImpulse, UpdatedBehaviourImpulse, DeletedBehaviourImpulse, CreateCodeImpulse, UpdateCodeImpulse, \
    DeleteCodeImpulse, CreatedCodeImpulse, UpdatedCodeImpulse, DeletedCodeImpulse, CreateEdgeImpulse, \
    DeleteEdgeImpulse, CreatedEdgeImpulse, DeletedEdgeImpulse, BehaviourExecutionRes, BehaviourExecutionReq, \
    BehaviourExecutionDetailReq, BehaviourExecutionDetailRes, \
    CreatedSkillBuildJobImpulseRes, CreatedSkillBuildJobImpulseReq, \
    CancelSkillBuildJobImpulse, CreateSkillBuildJobImpulse, CanceledSkillBuildJobImpulseRes, \
    CanceledSkillBuildJobImpulseReq, \
    PracticeReq, PracticeRes, SkillBuildJobRes, SkillBuildJobReq

Uuid = str

class ISensorFunction(ABC):

    @abstractmethod
    def retrieve(self, config: Callable[[RetrievalReq], None]) -> Observable[RetrievalRes]:
        pass

    @abstractmethod
    def retrieve_experience(self, config: Callable[[ExperienceReq], None]) -> Observable[ExperienceRes]:
        pass

    @abstractmethod
    def retrieve_knowledge(self, config: Callable[[KnowledgeReq], None]) -> Observable[KnowledgeRes]:
        pass

    @abstractmethod
    def retrieve_edges(self, source: Uuid, config: Callable[[EdgeReq], None], limit: int, offset: int) -> Observable[EdgeRes]:
        pass

    @abstractmethod
    def retrieve_edge(self, source: Uuid, edgeId: Uuid, config: Callable[[EdgeReq], None]) -> Observable[EdgeRes]:
        pass

    @abstractmethod
    def retrieve_identities(self, config: Callable[[IdentityReq], None], limit: int, offset: int) -> Observable[IdentityRes]:
        pass

    @abstractmethod
    def retrieve_identity(self, identity_id: Uuid, config: Callable[[IdentityReq], None]) -> Observable[IdentityRes]:
        pass

    @abstractmethod
    def retrieve_intents(self, identityId: Uuid, config: Callable[[IntentReq], None], limit: int, offset: int) -> Observable[IntentRes]:
        pass

    @abstractmethod
    def retrieve_intent(self, identity_id: Uuid, reference: Uuid, config: Callable[[IntentReq], None]) -> Observable[IntentRes]:
        pass

    @abstractmethod
    def retrieve_prompts(self, identity_id: Uuid, config: Callable[[PromptReq], None], limit: int, offset: int) -> Observable[PromptRes]:
        pass

    @abstractmethod
    def retrieve_prompt(self, identity_id: Uuid, reference: Uuid,  config: Callable[[PromptReq], None]) -> Observable[PromptRes]:
        pass

    @abstractmethod
    def retrieve_statements(self, identity_id: Uuid, config: Callable[[StatementReq], None], limit: int, offset: int) -> Observable[StatementRes]:
        pass

    @abstractmethod
    def retrieve_statement(self, identity_id: Uuid, reference: Uuid, config: Callable[[StatementReq], None]) -> Observable[StatementRes]:
        pass

    @abstractmethod
    def retrieve_fulfilments(self, identity_id: Uuid, config: Callable[[FulfilmentReq], None], limit: int, offset: int) -> Observable[FulfilmentRes]:
        pass

    @abstractmethod
    def retrieve_fulfilment(self, identity_id: Uuid, reference: Uuid, config: Callable[[FulfilmentReq], None]) -> Observable[FulfilmentRes]:
        pass

    @abstractmethod
    def retrieve_codes(self, identity_id: Uuid, config: Callable[[CodeReq], None], limit: int, offset: int) -> Observable[CodeRes]:
        pass

    @abstractmethod
    def retrieve_code(self, identity_id: Uuid, reference: Uuid, config: Callable[[CodeReq], None]) -> Observable[CodeRes]:
        pass

    @abstractmethod
    def retrieve_behaviours(self, identity_id: Uuid, config: Callable[[BehaviourReq], None], limit: int, offset: int) -> Observable[BehaviourRes]:
        pass

    @abstractmethod
    def retrieve_behaviour(self, identity_id: Uuid, reference: Uuid, config: Callable[[BehaviourReq], None]) -> Observable[BehaviourRes]:
        pass

    @abstractmethod
    def retrieve_behaviour_execution(self, identity_id: Uuid, process_instance_id: Uuid, config: Callable[[BehaviourExecutionDetailReq], None]) \
            -> Observable[BehaviourExecutionDetailRes]:
        pass

    @abstractmethod
    def retrieve_behaviour_executions(self, identity_id: Uuid, config: Callable[[BehaviourExecutionReq], None], limit: int = None, offset: int = None, startDate: str = None, endDate: str = None) \
            -> Observable[BehaviourExecutionRes]:
        pass

    @abstractmethod
    def introspect(self, config: Callable[[IntrospectionReq], None]) -> Observable[IntrospectionRes]:
        pass

    @abstractmethod
    def preserve(self, config: Callable[[PreservationReq], None]) -> Observable[PreservationRes]:
        pass

    @abstractmethod
    def preserve_create_identities(self, impulses: List[CreateIdentityImpulse]) -> Observable[CreatedIdentityImpulse]:
        pass

    @abstractmethod
    def preserve_update_identities(self, impulses: List[UpdateIdentityImpulse]) -> Observable[UpdatedIdentityImpulse]:
        pass

    @abstractmethod
    def preserve_delete_identities(self, impulses: List[DeleteIdentityImpulse]) -> Observable[DeletedIdentityImpulse]:
        pass

    @abstractmethod
    def preserve_create_intents(self, impulses: List[CreateIntentImpulse]) -> Observable[CreatedIntentImpulse]:
        pass

    @abstractmethod
    def preserve_update_intents(self, impulses: List[UpdateIntentImpulse]) -> Observable[UpdatedIntentImpulse]:
        pass

    @abstractmethod
    def preserve_delete_intents(self, impulses: List[DeleteIntentImpulse]) -> Observable[DeletedIntentImpulse]:
        pass

    @abstractmethod
    def preserve_create_prompts(self, impulses: List[CreatePromptImpulse]) -> Observable[CreatedPromptImpulse]:
        pass

    @abstractmethod
    def preserve_update_prompts(self, impulses: List[UpdatePromptImpulse]) -> Observable[UpdatedPromptImpulse]:
        pass

    @abstractmethod
    def preserve_delete_prompts(self, impulses: List[DeletePromptImpulse]) -> Observable[DeletedPromptImpulse]:
        pass

    @abstractmethod
    def preserve_create_statements(self, impulses: List[CreateStatementImpulse]) -> Observable[CreatedStatementImpulse]:
        pass

    @abstractmethod
    def preserve_update_statements(self, impulses: List[UpdateStatementImpulse]) -> Observable[UpdatedStatementImpulse]:
        pass

    @abstractmethod
    def preserve_delete_statements(self, impulses: List[DeleteStatementImpulse]) -> Observable[DeletedStatementImpulse]:
        pass

    @abstractmethod
    def preserve_create_fulfilments(self, impulses: List[CreateFulfilmentImpulse]) -> Observable[CreatedFulfilmentImpulse]:
        pass

    @abstractmethod
    def preserve_update_fulfilments(self, impulses: List[UpdateFulfilmentImpulse]) -> Observable[UpdatedFulfilmentImpulse]:
        pass

    @abstractmethod
    def preserve_delete_fulfilments(self, impulses: List[DeleteFulfilmentImpulse]) -> Observable[DeletedFulfilmentImpulse]:
        pass

    @abstractmethod
    def preserve_create_behaviours(self, impulses: List[CreateBehaviourImpulse]) -> Observable[CreatedBehaviourImpulse]:
        pass

    @abstractmethod
    def preserve_update_behaviours(self, impulses: List[UpdateBehaviourImpulse]) -> Observable[UpdatedBehaviourImpulse]:
        pass

    @abstractmethod
    def preserve_delete_behaviours(self, impulses: List[DeleteBehaviourImpulse]) -> Observable[DeletedBehaviourImpulse]:
        pass

    @abstractmethod
    def preserve_create_codes(self, impulses: List[CreateCodeImpulse]) -> Observable[CreatedCodeImpulse]:
        pass

    @abstractmethod
    def preserve_update_codes(self, impulses: List[UpdateCodeImpulse]) -> Observable[UpdatedCodeImpulse]:
        pass

    @abstractmethod
    def preserve_delete_codes(self, impulses: List[DeleteCodeImpulse]) -> Observable[DeletedCodeImpulse]:
        pass

    @abstractmethod
    def preserve_create_edges(self, impulses: List[CreateEdgeImpulse]) -> Observable[CreatedEdgeImpulse]:
        pass

    @abstractmethod
    def preserve_delete_edges(self, impulses: List[DeleteEdgeImpulse]) -> Observable[DeletedEdgeImpulse]:
        pass

    @abstractmethod
    def perceive(self, config: Callable[[PerceptionReq], None]) -> Observable[PerceptionRes]:
        pass

    @abstractmethod
    def perceive_action(self, impulse: PerceiveActionImpulse) -> Observable[PerceivedImpulse]:
        pass

    @abstractmethod
    def perceive_data(self, impulse: PerceiveDataImpulse) -> Observable[PerceivedImpulse]:
        pass

    @abstractmethod
    def practice(self, config: Callable[[PracticeReq], None]) -> Observable[PracticeRes]:
        pass

    @abstractmethod
    def practice_build(self, impulse: CreateSkillBuildJobImpulse, config: Union[Callable[[CreatedSkillBuildJobImpulseReq], None], None] = None) -> Observable[CreatedSkillBuildJobImpulseRes]:
        pass

    @abstractmethod
    def practice_cancel(self, impulse: CancelSkillBuildJobImpulse, config: Union[Callable[[CanceledSkillBuildJobImpulseReq], None], None] = None) -> Observable[CanceledSkillBuildJobImpulseRes]:
        pass

    @abstractmethod
    def introspect_build_jobs(self, tenant_id: Uuid, config: Union[Callable[[SkillBuildJobReq], None], None] = None) -> Observable[SkillBuildJobRes]:
        pass